### PR TITLE
Implementation API Queue

### PR DESCRIPTION
**Pull request by sjuarezgx**
Issue:96198

### CODE REVIEW

**Comment by ggallotti** [comment]
Nice! 👏 
Could we add UnitTesting, like the Storage Units Tests?

File: dotnet/src/dotnetcore/Providers/Messaging/GXQueue/Queue.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		public const int WAIT_TIMEOUT = 10;
+		public const int VISIBILITY_TIMEOUT = 60;
+
+		int GetQueueLength(out bool success);
+		MessageQueueResult SendMessage(SimpleQueueMessage simpleQueueMessage, out bool success);
+		MessageQueueResult SendMessage(SimpleQueueMessage simpleQueueMessage, MessageQueueOptions messageQueueOptions, out bool success);
+		IList<MessageQueueResult> SendMessages(IList<SimpleQueueMessage> simpleQueueMessages, MessageQueueOptions messageQueueOptions, out bool success);
+		IList<SimpleQueueMessage> GetMessages(out bool success);
+		IList<SimpleQueueMessage> GetMessages(MessageQueueOptions messageQueueOptions, out bool success);
+		MessageQueueResult DeleteMessage(out bool success);
```
Comment (ggallotti) [inline dotnet/src/dotnetcore/Providers/Messaging/GXQueue/Queue.cs:21]: Why this method? In the Spec the DeleteMessage receives a MessageId. 

This method should be called "Dequeue". It means take the first and remove it.

File: dotnet/src/dotnetcore/Providers/Messaging/GXAmazonSQS/AWSQueue.cs [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+				messageHandleId = new List<string>{ simpleQueueMessages[0].MessageHandleId };
+				messageQueueResults = RemoveMessages(messageHandleId, out bool operationOK);
+				if ((operationOK) && (messageQueueResults != null))
+					messageQueueResult = messageQueueResults[0];
+				success= true;
+			}
+			return messageQueueResult;
+		}
+
+		public IList<MessageQueueResult> DeleteMessages(List<string> messageHandleId, MessageQueueOptions messageQueueOptions, out bool success)
```
Comment (ggallotti) [inline dotnet/src/dotnetcore/Providers/Messaging/GXAmazonSQS/AWSQueue.cs:95]: Why DeleteMessages receives MessageQueueOptions? Seems to not be needed

File: dotnet/src/dotnetcore/Providers/Messaging/GXQueue/Queue.cs (hunk 1 above)
Comment (sjuarezgx) [reply dotnet/src/dotnetcore/Providers/Messaging/GXQueue/Queue.cs:21]: Yes, it isn't in the spec, but I found it usefull to dequeue. I can rename it, ok?

File: dotnet/src/dotnetcore/Providers/Messaging/GXAmazonSQS/AWSQueue.cs (hunk 2 above)
Comment (sjuarezgx) [reply dotnet/src/dotnetcore/Providers/Messaging/GXAmazonSQS/AWSQueue.cs:95]: In the case of Azure it is needed.

File: dotnet/src/dotnetcore/Providers/Messaging/GXQueue/Queue.cs (hunk 1 above)
Comment (ggallotti) [reply dotnet/src/dotnetcore/Providers/Messaging/GXQueue/Queue.cs:21]: ok!

File: dotnet/src/dotnetcore/Providers/Messaging/GXQueue/SimpleMessageQueue.cs [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+			catch (Exception ex)
+			{
+				QueueErrorMessagesSetup(ex, out errorMessages);
+				GXLogging.Error(logger, ex);
+				success = false;
+			}
+			return TransformMessageQueueResult(messageQueueResult);
+		}
+
+		public IList<GxUserType> DeleteMessages(List<string> messageHandleId, GxUserType messageQueueOptions, out GXBaseCollection<SdtMessages_Message> errorMessages, out bool success)
```
Comment (ggallotti) [inline dotnet/src/dotnetcore/Providers/Messaging/GXQueue/SimpleMessageQueue.cs:106]: messageQueueOptions parameter should be removed.

File: dotnet/src/dotnetcore/Providers/Messaging/GXAzureQueue/AzureQueue.cs [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+				}
+			}
+			return messageQueueResult;
+		}
+
+		/// <summary>
+		/// Deletes permanently the messages given on the list.
+		/// </summary>
+		
+		public IList<MessageQueueResult> DeleteMessages(List<string> messageHandleId, MessageQueueOptions messageQueueOptions,out bool success)
```
Comment (ggallotti) [inline dotnet/src/dotnetcore/Providers/Messaging/GXAzureQueue/AzureQueue.cs:145]: Delete Message should not receive messageQueueOptions. The user just want to delete a message.

File: dotnet/src/dotnetcore/Providers/Messaging/GXAzureQueue/AzureQueue.cs [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+		{
+			success = false;
+			IList<MessageQueueResult> messageQueueResults = new List<MessageQueueResult>();
+			if (_queueClient is QueueClient && _queueClient.Exists())
+			{
+				QueueProperties properties = _queueClient.GetProperties();
+
+				QueueMessage[] receivedMessages = _queueClient.ReceiveMessages(messageQueueOptions.MaxNumberOfMessages);
```
Comment (ggallotti) [inline dotnet/src/dotnetcore/Providers/Messaging/GXAzureQueue/AzureQueue.cs:153]: Does  not exists a direct way to delete message just with their messageId? Seems unreliable to need to consume all the queue in order to be able to delete. The only case is see for this is FIFO Queues.

Comment (sjuarezgx) [reply dotnet/src/dotnetcore/Providers/Messaging/GXAzureQueue/AzureQueue.cs:153]: No, I need the PopReceipt of the message.

**Review by claudiamurialdo — APPROVED** [review]

File: dotnet/src/dotnetcore/Providers/Messaging/GXQueue/SimpleMessageQueue.cs [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+				}
+				else
+				{
+					GXLogging.Error(logger, "(GXQueue1002)Queue Error", ex);
+					GXUtil.ErrorToMessages("GXQueue1002", ex, errorMessages);
+				}
+			}
+		}
+	}
+	public class ServiceFactory
```
[the remark this answers is not in the document]
Comment (sjuarezgx) [reply dotnet/src/dotnetcore/Providers/Messaging/GXQueue/SimpleMessageQueue.cs]: Yes, I had already fixed that. Let me see what commit wasn't merged..

**Review by claudiamurialdo — APPROVED** [review]

**Pull request merged**